pydemolib/nessiedemo/flink.py

```diff
@@ -0,0 +1,168 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020 Dremio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""NessieDemoFlink handles setting up Flink and Iceberg related objects."""
+
+import os
+import re
+import shutil
+from types import TracebackType
+from typing import Any, Tuple, TypeVar
+
+from pyflink.datastream import StreamExecutionEnvironment
+
+from .demo import _Util, NessieDemo
+
+T = TypeVar("T", bound="NessieDemoFlink")
+
+
+class NessieDemoFlink:
+    """`NessieDemoFlink` is a helper class for Flink in Nessie-Demos."""
+
+    __demo: NessieDemo
+
+    __env: StreamExecutionEnvironment
+    __jvm: Any
+
+    def __init__(self: T, demo: NessieDemo) -> None:
+        """Creates a `NessieDemoFlink` instance for respectively using the given `NessieDemo` instance."""
+        self.__demo = demo
+
+        if "hadoop" in self.__demo._get_versions_dict() and "tarball" in self.__demo._get_versions_dict()["hadoop"]:
+            hadoop_url = self.__demo._get_versions_dict()["hadoop"]["tarball"]
+            # derive directory name inside the tarball from the URL
+            m = re.match(".*[/]([a-zA-Z0-9-.]+).tar.gz", hadoop_url)
+            if not m:
+                raise Exception("Invalid Hadoop download URL {}".format(hadoop_url))
+            dir_name = m.group(1)
+            hadoop_dir = self.__demo._asset_dir(dir_name)
+            if not os.path.exists(hadoop_dir):
+                tgz = self.__demo._asset_dir("{}tar.gz".format(dir_name))
+                if not os.path.exists(tgz):
+                    _Util.wget(hadoop_url, tgz)
+                _Util.exec_fail(["tar", "-x", "-z", "-C", os.path.abspath(os.path.join(hadoop_dir, "..")), "-f", tgz])
+        else:
+            raise Exception("configuration does not define hadoop.tarball. Unable to find Hadoop.")
+
+        print("Using Hadoop in {}".format(hadoop_dir))
+
+        os.environ["HADOOP_HOME"] = hadoop_dir
+        self.__copy_all_hadoop_jars_to_pyflink()
+
+    def __copy_all_hadoop_jars_to_pyflink(self: T) -> T:
+        jar_files = []
+
+        for root, _, files in os.walk(os.getenv("HADOOP_HOME")):
+            for file in files:
+                if file.endswith(".jar"):
+                    jar_files.append(os.path.join(root, file))
+
+        import site
+        pyflink_lib_dir = os.path.join(site.getsitepackages()[0], "pyflink", "lib")
```

Comment:
I'd prefer a function in `_Util` to find the directory for a python package, unifying this one with the similar one in `NessieDemo.__init__`.

Reply:
ok fixed